CMSgov/mint-app
# Easi 1996/plan milestones status

# EASI-1996

## Changes and Description

- Add db:seed data for milestones
- Add missing fields in DB tables and models to match up to FIGMA
- Add calculated status field like in plan basics



## How to test this change

`scripts/dev db:seed` --> Ensure that there is seeded data

Create Model Plan to test against
```gql
mutation createModelPlan {
  createModelPlan(input: {modelName: "My Custom Model", id: "9b5db64b-8d42-4b91-a601-d6f9732b5298", status: PLAN_DRAFT}) {
    id
  }
}
```
Create Plan Milestones
```gql
mutation createPlanMilestones #($input: PlanMilestonesInput!)
 {
    createPlanMilestones (input: 
    {
        id: "71a6ddaa-f365-4089-8c32-e4b04b2ce6f1"
        modelPlanID: "9b5db64b-8d42-4b91-a601-d6f9732b5298"
        clearanceStarts: "2022-04-02T15:04:05.999999999Z"
        clearanceEnds: "2022-04-02T15:04:05.999999999Z"
        announced: "2022-04-02T15:04:05.999999999Z"
        applicationsStart: "2022-04-02T15:04:05.999999999Z"
        applicationsEnd: "2022-04-02T15:04:05.999999999Z"
        performancePeriodStarts: "2022-04-02T15:04:05.999999999Z"
        performancePeriodEnds: "2022-04-02T15:04:05.999999999Z"
        wrapUpEnds: "2022-04-02T15:04:05.999999999Z"
        highLevelNote: "All Milestones are currently just best estimates"
        phasedIn: false
        phasedInNote: "This can't be phased in, it is all or nothing"

        # status: READY
    }
    
    ) {
        id
        modelPlanID
        completeICIP
        clearanceStarts
        clearanceEnds
        announced
        applicationsStart
        applicationsEnd
        performancePeriodStarts
        performancePeriodEnds
        wrapUpEnds
        highLevelNote
        phasedIn
        phasedInNote
        createdBy
        createdDts
        modifiedBy
        modifiedDts
        status
    }
}
```
Update Plan Milestones
```gql
mutation updatePlanMilestones #($input: PlanMilestonesInput!) 
{
    updatePlanMilestones (input: 
        {
        id: "71a6ddaa-f365-4089-8c32-e4b04b2ce6f1"
        modelPlanID: "9b5db64b-8d42-4b91-a601-d6f9732b5298"
        completeICIP:"2023-04-02T15:04:05.999999999Z"
        clearanceStarts: "2022-04-02T15:04:05.999999999Z"
        clearanceEnds: "2022-04-02T15:04:05.999999999Z"
        announced: "2022-04-02T15:04:05.999999999Z"
        applicationsStart: "2022-04-02T15:04:05.999999999Z"
        applicationsEnd: "2022-04-02T15:04:05.999999999Z"
        performancePeriodStarts: "2022-04-02T15:04:05.999999999Z"
        performancePeriodEnds: "2022-04-02T15:04:05.999999999Z"
        wrapUpEnds: "2022-04-02T15:04:05.999999999Z"
        highLevelNote: "All Milestones are currently just best estimates"
        # phasedIn: true
        phasedInNote: "It turns out this can be phased in over time"

        # status: READY
    }) {
        id
        modelPlanID
        completeICIP
        clearanceStarts
        clearanceEnds
        announced
        applicationsStart
        applicationsEnd
        performancePeriodStarts
        performancePeriodEnds
        wrapUpEnds
        highLevelNote
        phasedIn
        phasedInNote
        createdBy
        createdDts
        modifiedBy
        modifiedDts
        status
    }
}
```

Get Model Plan with Milestones
```gql
query ModelPlan{
    modelPlan( id: "9b5db64b-8d42-4b91-a601-d6f9732b5298"){
#   modelPlan( id: "6e224030-09d5-46f7-ad04-4bb851b36eab") {
    id
    createdBy
    # cmsCenters
    # cmsOther
    # cmmiGroups
    status
    archived
    milestones {
                id
        modelPlanID
        completeICIP
        clearanceStarts
        clearanceEnds
        announced
        applicationsStart
        applicationsEnd
        performancePeriodStarts
        performancePeriodEnds
        wrapUpEnds
        highLevelNote
        phasedIn
        phasedInNote
    }
    # collaborators {
    #   id
    #   modelPlanID
    #   euaUserID
    #   teamRole
    #   fullName
    # }
    # discussions {
    #     id
    #     modelPlanID
    #     content
    #     status
    #     createdBy
    #     createdDts
    #     modifiedBy
    #     modifiedDts
    #     replies {
    #         id
    #         discussionID
    #         content
    #         resolution
    #         createdBy
    #         createdDts
    #         modifiedBy
    #         modifiedDts

    #         }
    # }
    basics{
        id
        modelPlanID
        modelType
        problem
        goal
        testInventions
        note
        createdBy
        createdDts
        modifiedBy
        modifiedDts
        status        
    }
  }
}
```


## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
